extensions/eclipse-che-theia-plugin-ext/src/browser/che-github-main.ts

```diff
@@ -0,0 +1,105 @@
+
+import { CheGithubMain } from '../common/che-protocol';
+import { interfaces } from 'inversify';
+import { EnvVariablesServer } from '@theia/core/lib/common/env-variables';
+import axios, { AxiosInstance } from 'axios';
+
+export class CheGithubMainImpl implements CheGithubMain {
+    private readonly envVariableServer: EnvVariablesServer;
+    private axiosInstance: AxiosInstance = axios;
+    private apiUrl: string;
+    private token: string | undefined;
+
+    constructor(container: interfaces.Container) {
+        this.envVariableServer = container.get(EnvVariablesServer);
+        this.envVariableServer.getValue('CHE_API').then(variable => {
+            if (variable && variable.value) {
+                this.apiUrl = variable.value;
+            }
+        });
+    }
+
+    async $uploadPublicSshKey(publicKey: string): Promise<void> {
+        await this.fetchToken();
+        await this.axiosInstance.post('https://api.github.com/user/keys?access_token=' + this.token, {
+            title: 'che-theia',
+            key: publicKey
+        });
+    }
+
+    private async fetchToken(): Promise<void> {
+        if (!this.token) {
+            await this.updateToken();
+        } else {
+            try {
+                await this.axiosInstance.get('https://api.github.com/user?access_token=' + this.token);
+            } catch (e) {
+                await this.updateToken();
+            }
+        }
+    }
+
+    private async updateToken(): Promise<void> {
+        this.token = await this.getToken();
+        if (!this.token) {
+            await this.authenticate();
+            this.token = await this.getToken();
+        }
+    }
+
+    private authenticate(): Promise<void> {
+        return new Promise(async resolve => {
+            const redirectUrl = window.location.href;
+            const url = `${this.apiUrl}/oauth/authenticate?oauth_provider=github&userId=${await this.getUserId()}` +
+                `&scope=write:public_key&redirect_after_login=${redirectUrl}`;
+            const popupWindow = window.open(url, 'popup');
+            const popup_close_handler = async () => {
+                if (!popupWindow || popupWindow.closed) {
+                    if (popupCloseHandlerIntervalId) {
+                        window.clearInterval(popupCloseHandlerIntervalId);
+                    }
+                } else {
+                    let href;
+                    try {
+                        href = popupWindow.location.href;
+                    } catch (error) {
+                    }
+                    if (href) {
+                        if (href === redirectUrl) {
+                            popupWindow.close();
+                            if (popupCloseHandlerIntervalId) {
+                                window.clearInterval(popupCloseHandlerIntervalId);
+                            }
+                            resolve();
+                        }
+                    }
+                }
+            };
+
```

Comment:
there is no reject of the promise in case of error ?

Reply:
Added `reject()` when the authorization window closed and not authorized.